pkg/client/block/client.go

```diff
@@ -36,42 +39,148 @@ func NewBlockClient(
 	ctx context.Context,
 	deps depinject.Config,
 ) (client.BlockClient, error) {
-	client, err := events.NewEventsReplayClient[client.Block](
+	ctx, close := context.WithCancel(ctx)
+
+	eventsReplayClient, err := events.NewEventsReplayClient[client.Block](
 		ctx,
 		deps,
 		committedBlocksQuery,
 		UnmarshalNewBlock,
 		defaultBlocksReplayLimit,
 	)
 	if err != nil {
+		close()
+		return nil, err
+	}
+
+	// latestBlockPublishCh is the channel that notifies the latestBlockReplayObs of a
+	// new block, whether it comes from a direct query or an event subscription query.
+	latestBlockReplayObs, latestBlockPublishCh := channel.NewReplayObservable[client.Block](ctx, 10)
+	blockClient := &blockReplayClient{
+		eventsReplayClient:   eventsReplayClient,
+		latestBlockReplayObs: latestBlockReplayObs,
+		close:                close,
+	}
+
+	if err := depinject.Inject(deps, &blockClient.onStartQueryClient); err != nil {
+		return nil, err
+	}
+
+	blockClient.asyncForwardBlockEvent(ctx, latestBlockPublishCh)
+
+	if err := blockClient.getInitialBlock(ctx, latestBlockPublishCh); err != nil {
 		return nil, err
 	}
-	return &blockClient{eventsReplayClient: client}, nil
+
+	return blockClient, nil
 }
 
-// blockClient is a wrapper around an EventsReplayClient that implements the
+// blockReplayClient is a wrapper around an EventsReplayClient that implements the
 // BlockClient interface for use with cosmos-sdk networks.
-type blockClient struct {
+type blockReplayClient struct {
+	// onStartQueryClient is the RPC client that is used to query for the initial block
+	// upon blockReplayClient construction. The result of this query is only used if it
+	// returns before the eventsReplayClient receives its first event.
+	onStartQueryClient cometclient.CometRPC
+
 	// eventsReplayClient is the underlying EventsReplayClient that is used to
 	// subscribe to new committed block events. It uses both the Block type
 	// and the BlockReplayObservable type as its generic types.
 	// These enable the EventsReplayClient to correctly map the raw event bytes
 	// to Block objects and to correctly return a BlockReplayObservable
 	eventsReplayClient client.EventsReplayClient[client.Block]
+
+	// latestBlockReplayObs is a replay observable that combines blocks observed by
+	// the block query client & the events replay client. It is the "canonical"
+	// source of block notifications for blockClient.
+	latestBlockReplayObs observable.ReplayObservable[client.Block]
+
+	// close is a function that cancels the context of the blockClient.
```

Comment:
s/blockClient/blockReplayClient

Please make sure there are no other lingering instances of this as well.